### PR TITLE
Login endpoint on server

### PR DESCRIPTION
**Pull request by hendrikebbers**
Provides a login endpoint on the server (servlet based) in the security module.
The endpoint is a proxy to the Keycloak server that can be used for login and token refresh. By doing so the client do not need to know / access the KeyCloak server.

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/819)

### CODE REVIEW

File: platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/dp/impl/security/KeycloakSecurity.java [hunk 1]
```diff
@@ -42,56 +46,63 @@
...
 
-    public KeycloakSecurity(final String authEndpoint, final String realmName, final String appName, final ExecutorService executor) {
-        this.appName = Assert.requireNonBlank(appName, "appName");
-        this.authEndpoint = Assert.requireNonBlank(authEndpoint, "authEndpoint");
-        this.realmName = Assert.requireNonBlank(realmName, "realmName");
-        this.executor = Assert.requireNonNull(executor, "executor");
+    public KeycloakSecurity(final ClientConfiguration configuration) {
+        Assert.requireNonNull(configuration, "configuration");
+
+        this.appName =configuration.getProperty(APPLICATION_PROPERTY_NAME, APPLICATION_PROPERTY_DEFAULT_VALUE);
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/dp/impl/security/KeycloakSecurity.java]: Add `blank` after equality sign.

File: platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/dp/impl/security/KeycloakSecurity.java [hunk 2]
```diff
@@ -42,56 +46,63 @@
...
         }
-        final byte[] rawInput = byteArrayOutputStream.toByteArray();
-        final String input = new String(rawInput);
+    }
 
+    private void receiveTokenFromKeycloak(final String content) throws IOException, URISyntaxException {
+        Assert.requireNonNull(content, "content");
+        LOG.debug("receiving new token from keycloak server");
+        final HttpClientConnection clientConnection = createConnection();
+        clientConnection.addRequestHeader("Content-Type", "application/x-www-form-urlencoded");
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/dp/impl/security/KeycloakSecurity.java]: Is it possible to use constants for the MIME Type that are already defined somewhere?

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakTokenServlet.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            clientConnection.addRequestHeader("Content-Type", "application/x-www-form-urlencoded");
+            clientConnection.setDoOutput(true);
+            clientConnection.writeRequestContent(content);
+            final int responseCode = clientConnection.readResponseCode();
+            if(responseCode == SC_HTTP_UNAUTHORIZED) {
+                throw new DolphinRuntimeException("Invalid login!");
+            }
+            final byte[] responseContent = clientConnection.readResponseContent();
+            ConnectionUtils.writeContent(resp.getOutputStream(), responseContent);
+        } catch (Exception e) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakTokenServlet.java]: Add `final` keyword.

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/http/ConnectionUtils.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+
+import static com.canoo.dp.impl.platform.core.http.HttpHeaderConstants.CHARSET;
+
+public class ConnectionUtils {
+
+    private ConnectionUtils() {}
+
+    public static byte[] readContent(final InputStream inputStream) throws IOException {
```
Comment (giftkugel) [inline platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/http/ConnectionUtils.java]: Add null-check for `inputStream`.

File: platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/dp/impl/security/KeycloakSecurity.java (hunk 1 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/dp/impl/security/KeycloakSecurity.java]: done

File: platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/dp/impl/security/KeycloakSecurity.java (hunk 2 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-security-client/src/main/java/com/canoo/dp/impl/security/KeycloakSecurity.java]: done

File: platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakTokenServlet.java (hunk 3 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-security-server/src/main/java/com/canoo/dp/impl/server/security/KeycloakTokenServlet.java]: done

File: platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/http/ConnectionUtils.java (hunk 4 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-core/src/main/java/com/canoo/dp/impl/platform/core/http/ConnectionUtils.java]: done

**Pull request merged**